test/utils/oidc/testserver.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/tls"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	openIDWellKnownWebPath = "/.well-known/openid-configuration"
+	authWebPath            = "/auth"
+	tokenWebPath           = "/token"
+	jwksWebPath            = "/jwks"
+	userInfoWebPath        = "/user-info"
+)
+
+var (
+	ErrRefreshTokenExpired   = errors.New("refresh token is expired")
+	ErrClientSecretNotPassed = errors.New("client secret is not passed")
+	ErrBadClientSecret       = errors.New("client secret is bad")
+	ErrBadClientID           = errors.New("client ID is bad")
+)
+
+type TestServer struct {
+	httpServer   *httptest.Server
+	tokenHandler *MockTokenHandler
+	jwksHandler  *MockJWKsHandler
+	TearDownFn   func()
+}
+
+// JwksHandler is getter of JSON Web Key Sets handler
+func (ts *TestServer) JwksHandler() *MockJWKsHandler {
+	return ts.jwksHandler
+}
+
+// TokenHandler is getter of JWT token handler
+func (ts *TestServer) TokenHandler() *MockTokenHandler {
+	return ts.tokenHandler
+}
+
+// URL returns the public URL of server
+func (ts *TestServer) URL() string {
+	return ts.httpServer.URL
+}
+
+// TokenURL returns the public URL of JWT token endpoint
+func (ts *TestServer) TokenURL() string {
+	return ts.httpServer.URL + tokenWebPath
+}
+
+// BuildAndRunTestServer configures OIDC TLS server and its routing
+func BuildAndRunTestServer(t *testing.T, caPath, caKeyPath string) *TestServer {
+	t.Helper()
+
+	certContent, err := os.ReadFile(caPath)
+	require.NoError(t, err)
+	keyContent, err := os.ReadFile(caKeyPath)
+	require.NoError(t, err)
+
+	cert, err := tls.X509KeyPair(certContent, keyContent)
+	require.NoError(t, err)
+
+	mux := http.NewServeMux()
+	httpServer := httptest.NewUnstartedServer(mux)
+	httpServer.TLS = &tls.Config{
+		Certificates: []tls.Certificate{cert},
+	}
+	httpServer.StartTLS()
```

Comment:
this may be a candidate to leak, it seems the server is never stopped

Reply:
#115759 should have fixed the leak - rebase on top of that and see if it works now.  I have not had a chance to go through the other questions / comments / etc in this PR yet - hoping to do so later in the release cycle.  Thank you for all the hard work done so far!

Reply:
> https://github.com/kubernetes/kubernetes/pull/115759 should have fixed the leak

Yep, that helped, thanks a lot!

> Thank you for all the hard work done so far!

That's my pleasure:)